File: studio/policy.py
"""Role policy checks for builder mutations."""

from fastapi import HTTPException

ROLE_MATRIX = {
    "owner": frozenset({"read", "edit", "generate", "restore", "share", "publish"}),
    "editor": frozenset({"read", "edit", "generate", "restore"}),
    "viewer": frozenset({"read"}),
}


def effective_role(project, user_id):
    owner = project.get("ownerUserId") or "local-user"
    if user_id == owner:
        return "owner"
    for collab in project.get("collaborators") or []:
        if collab.get("userId") == user_id:
            return collab.get("role", "viewer")
    if project.get("visibility") == "org":
        return "editor"
    return "viewer" if user_id != owner else "owner"


def require_action(project, user_id, action):
    role = effective_role(project, user_id)
    allowed = ROLE_MATRIX.get(role, frozenset())
    if action not in allowed:
        raise HTTPException(
            status_code=403,
            detail={"code": "forbidden", "action": action, "role": role},
        )
    return role

File: src/server/policy.ts
import { HttpError } from "./api";

const ROLE_MATRIX: Record<string, Set<string>> = {
  owner: new Set(["read", "edit", "generate", "restore", "share", "publish", "delete"]),
  editor: new Set(["read", "edit", "generate", "restore"]),
  viewer: new Set(["read"]),
};

export function effectiveRole(
  project: Record<string, unknown>,
  userId: string,
): string {
  const owner = (project.ownerUserId as string | undefined) ?? "local-user";
  if (userId === owner) return "owner";
  // Pre-auth dev projects keep ownerUserId "local-user" until legacy adoption runs.
  if (owner === "local-user" && userId) return "owner";
  const collaborators = (project.collaborators as Array<Record<string, unknown>> | null) ?? [];
  for (const collab of collaborators) {
    if (collab.userId === userId) return (collab.role as string | undefined) ?? "viewer";
  }
  if (project.visibility === "org") return "editor";
  return userId !== owner ? "viewer" : "owner";
}

export function requireAction(
  project: Record<string, unknown>,
  userId: string,
  action: string,
): string {
  const role = effectiveRole(project, userId);
  const allowed = ROLE_MATRIX[role] ?? new Set<string>();
  if (!allowed.has(action)) {
    throw new HttpError(403, JSON.stringify({ code: "forbidden", action, role }));
  }
  return role;
}
